File: core/x/javaapi/api/src/main/java/viaduct/java/api/documents/FragmentFromAnnotation.java
package viaduct.java.api.documents;

import viaduct.java.api.annotations.GraphQLFragment;
import viaduct.java.api.types.CompositeOutput;

/**
 * Base class for reusable named GraphQL fragments.
 *
 * <p>Subclasses must be annotated with {@link GraphQLFragment}. The type parameter identifies the
 * GraphQL composite type selected by the fragment and is checked against the fragment's type
 * condition during tenant-module assembly.
 *
 * @param <T> the GraphQL composite output type selected by the fragment
 */
public abstract class FragmentFromAnnotation<T extends CompositeOutput> {
  /** Returns the fragment document declared by {@link GraphQLFragment}. */
  public final String getFragmentText() {
    GraphQLFragment annotation = getClass().getAnnotation(GraphQLFragment.class);
    if (annotation == null) {
      throw new IllegalStateException(
          getClass().getSimpleName() + " must be annotated with @GraphQLFragment");
    }
    return annotation.value();
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/documents/FragmentFromAnnotation.kt
package viaduct.api.documents

import viaduct.api.types.CompositeOutput
import viaduct.apiannotations.ExperimentalApi

/**
 * Abstract base class for named GraphQL fragment objects.
 *
 * Extend this class with a Kotlin singleton `object` and annotate it with [GraphQLFragment] to
 * declare a reusable GraphQL fragment.
 *
 * ```kotlin
 * @GraphQLFragment("""
 *   fragment UserCoreFields on User {
 *     id
 *     name
 *     email
 *   }
 * """)
 * object UserCoreFieldsFragment : FragmentFromAnnotation<User>()
 * ```
 *
 * Spread the fragment in resolver fragments:
 * ```kotlin
 * @Resolver(objectValueFragment = "fragment _ on Account { ...UserCoreFields }")
 * class AccountDisplayNameResolver : AccountResolvers.DisplayName() { ... }
 * ```
 *
 * or in subquery operations:
 * ```kotlin
 * @GraphQLOperation("query { viewer { ...UserCoreFields } }")
 * object ViewerQuery : QueryFromAnnotation()
 * val result = ctx.query(ViewerQuery)
 * ```
 *
 * The subclass must be a Kotlin singleton `object` annotated with [GraphQLFragment].
 * Named fragments are scoped to the tenant module in which they are declared.
 *
 * @param T the GraphQL object type on which this fragment selects fields — must be a composite
 *   output type (object, interface, or union).
 *
 * @see GraphQLFragment
 * @see [Resolver][viaduct.api.resolver.Resolver]
 */
@ExperimentalApi
abstract class FragmentFromAnnotation<T : CompositeOutput> {
    /**
     * The GraphQL fragment definition text declared in [@GraphQLFragment][GraphQLFragment].
     *
     * @throws IllegalStateException if the subclass is not annotated with [@GraphQLFragment][GraphQLFragment].
     */
    val fragmentText: String by lazy {
        this::class.annotations
            .filterIsInstance<GraphQLFragment>()
            .firstOrNull()
            ?.value
            ?: error("${this::class.simpleName} must be annotated with @GraphQLFragment")
    }
}
